mentat/parsers/parser.py

```diff
@@ -33,11 +49,276 @@ def interrupt_catcher(self):
     def get_system_prompt(self) -> str:
         pass
 
-    @abstractmethod
     async def stream_and_parse_llm_response(
         self,
         response: AsyncGenerator[Any, None],
         code_file_manager: CodeFileManager,
         config: ConfigManager,
     ) -> tuple[str, list[FileEdit]]:
+        """
+        This general parsing structure relies on the assumption that all formats require three types of lines:
+        1. 'conversation' lines, which are streamed as they come,
+        2. 'special' lines, that are never shown to the user and contain information such as the file_name
+        3. 'code' lines, which are the actual code written and are shown to the user in a special format
+        To make a parser that differs from these assumptions, make this functionality a subclass of Parser
+        """
```

Comment:
This function is 200 lines and nested pretty deep, which isn't ideal.

Doesn't have to be refactored in this PR but after you've made more parsing formats and are more sure that this generalizes how we want we can take some time to improve it.